core/src/main/java/org/pentaho/di/connections/utils/ConnectionDetailsUtils.java

```diff
@@ -0,0 +1,63 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2024 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.connections.utils;
+
+import org.pentaho.di.connections.ConnectionDetails;
+import org.pentaho.metastore.api.IMetaStore;
+import org.pentaho.metastore.api.exceptions.MetaStoreException;
+import org.pentaho.metastore.persist.MetaStoreFactory;
+import org.pentaho.metastore.stores.memory.MemoryMetaStore;
+
+import static org.pentaho.metastore.util.PentahoDefaults.NAMESPACE;
+
+public class ConnectionDetailsUtils {
+
+  private ConnectionDetailsUtils() {
+    // Static clas
+  }
+
+  /**
+   * Clones a given connection details instance.
+   * <p>
+   * Cloning is done by according to the meta store metadata from present meta store annotations in the connection
+   * details type. Essentially, behaving as a save followed by a load from an in-memory meta-store.
+   * As such any information not stored in the meta store is not preserved in the cloned instance.
+   * <p>
+   * The given connection details must have a name set or the operation fails.
+   *
+   * @param instance The connection details instance to clone.
+   * @return The cloned connection details.
+   */
+  public static <T extends ConnectionDetails> T cloneMeta( T instance ) throws MetaStoreException {
+    // Use a temporary in-memory meta store to save to and then load from.
+    IMetaStore metaStore = new MemoryMetaStore();
```

Comment:
I read the comment, but why was this route taken to use the class `MemoryMetaStore` in `ConnectionDetails#cloneMeta` ?
Why was a standard clone function not used or implemented?
Was it for expediency?  

If it was for expediency, I would consider this technical debt that should be tracked an addressed at a later date determined by PO/Managers.

From my research, the class `MemoryMetaStore.java` has not been updated with source code change in close to 10 years and wasn't in the recent MetaStore effort https://hv-eng.atlassian.net/browse/BACKLOG-37507.

My main concern is maintainability.

Reply:
The _fields_ of a `ConnectionDetails` instance that need to be encrypted and decrypted across the network are those marked with both the `@MetaStoreAttribute` and `@Encrypted` annotation.

The `ConnectionDetails` interface mixes other concerns such as opening and closing an SWT dialog (`openDialog` and `closeDialog`), and existing implementations typically store a field holding the currently open dialog. A dialog instance which is associated with the specific connection details instance and would not be a valid association if a blind clone approach would be used. Also, all implementations have this field as a private/package field.

The approach of using the generic, standard `Object#clone` method proved harder than it looked at first. Either as a new default method of the interface, or as a separate utility. The `clone` method is protected and would require existing classes to be changed to access its functionality. While there are base classes used by our implementations, there's no assurance other implementations do not exist out there. Alternatively, we'd need to use reflection to call the `clone` method. Still, we'd likely be knowing more than we should by using it this way.
The `clone` method is seen by many, these days, as a having a flawed design.

If we would be throwing our own clone way, then we'd need cooperation from all existing implementations, effectively breaking compatibility.

Using the meta store utilities, however, provides a straightforward generic solution to the problem (that solves all cases and does not require any code per connection type), given it already deals with identifying the parts of the connection details which are relevant to persist. The same parts which are relevant to persist are the ones relevant to communicate to the web client.

The meta-store also additionally already handles deep "cloning" / persistence of the meta-store properties, a use case that the native clone implementation would not address.

The `MemoryMetaStore` class is the base class of:
- `PurRepositoryMetaStore`, used to hold the repository in memory, when connected to a Pentaho Server
- `KettleDatabaseRepositoryMetaStore`, used by `KettleDatabaseRepository`, which I assume is the repository used when not connected to a Pentaho server.
- `SharedObjectsMetaStore` which is deprecated

So I think it is safe to conclude that `MemoryMetaStore` is up to date and will be maintained for other purposes.

Finally, I took the idea from something similar that the `ConnectionManager` does, when its local connection details instance becomes modified by encryption before saving: https://github.com/pentaho/pentaho-kettle/blob/e8140d64f33cfe6b158d7c92972c78f59625a12e/core/src/main/java/org/pentaho/di/connections/ConnectionManager.java#L232-L237.

Reply:
thanks for the explanation